controller-server/src/main/java/com/yahoo/vespa/hosted/controller/notification/NotificationContentFormatter.java

```diff
@@ -0,0 +1,209 @@
+            case perf:
+                return link.apply(environment);
+            default:
+                return link.apply(Environment.prod);
+        }
+    }
+
+    private String jobText(NotificationSource source) {
+        var jobType = expect(source.jobType());
+        var zone = jobType.zone();
+        var runNumber = source.runNumber().orElseThrow(MissingOptionalException::new);
+        switch (zone.environment().value()) {
+            case "production":
+                return Text.format("Deployment job #%d to %s", runNumber, zone.region());
+            case "test":
+                return Text.format("Test job #%d to %s", runNumber, zone.region());
+            case "dev":
+            case "perf":
+                return Text.format("Deployment job #%d to %s.%s", runNumber, zone.environment().value(), zone.region().value());
+        }
+        switch (jobType.jobName()) {
+            case "system-test":
+            case "staging-test":
+        }
+        return Text.format("%s #%d", jobType.jobName(), runNumber);
+    }
+
+    private String levelText(Notification.Level level, int count) {
+        switch (level) {
+            case error:
+                return "failed";
+            case warning:
+                return count > 1 ? Text.format("%d warnings", count) : "a warning";
+            default:
+                return count > 1 ? Text.format("%d messages", count) : "a message";
+        }
+    }
+
+    private String clusterInfo(NotificationSource source) {
+        var application = expect(source.application());
+        var instance = expect(source.instance());
+        var zone = expect(source.zoneId());
+        var clusterId = expect(source.clusterId());
+        return Text.format("Cluster %s in %s.%s for %s.%s",
+                clusterId.value(),
+                zone.environment(), zone.region(),
+                application, instance);
+    }
+
+    private URI url(Notification n) {
+        var source = n.source();
+        if (source.application().isPresent()) {
+            if (source.instance().isPresent()) {
+                if (source.jobType().isPresent() && source.runNumber().isPresent()) {
+                    return zoneRegistry.dashboardUrl(
+                            new RunId(ApplicationId.from(source.tenant(),
+                                    source.application().get(),
+                                    source.instance().get()),
+                                    source.jobType().get(),
+                                    source.runNumber().getAsLong()));
+                }
+                return zoneRegistry.dashboardUrl(ApplicationId.from(source.tenant(), source.application().get(), source.instance().get()));
+            }
+            return zoneRegistry.dashboardUrl(source.tenant(), source.application().get());
+        }
+        return zoneRegistry.dashboardUrl(source.tenant());
+    }
+
+
+    private static <T> T expect(Optional<T> optional) {
```

Comment:
Our convention is to name methods that verify some constraint `requireXXX`, e.g. `requirePresent`.

For this particular case I would suggest `requirePresent(Optional<T> optional, String fieldName)` and include `fieldName` in the exception message so that we can know which field is missing:

`requirePresent(source.application(), "application")`